schemachanger: port over MR zone config helpers

annrpom opened the pull request:
Some precursor work to remove RBR fallbacks for ADD COLUMN.

Epic: [CRDB-31282](https://cockroachlabs.atlassian.net/browse/CRDB-31282)
Informs: https://github.com/cockroachdb/cockroach/issues/80545

Release note: None

spilchen reviewed (commented):
> Reviewed 6 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141386)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @annrpom and @rafiss)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 1260 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141386#-OJdp1Kd6r9r-CPyfF2X:-OJdp1Ke3KaUlg5VCV5q:bl7nww8) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d6d787c7f43c40fd0b1825a93cfd0141f5188c8/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L1260)):*
> ```Go
> 	return b.QueryByID(tableID).FilterTemporaryIndex().
> 		Filter(func(_ scpb.Status, _ scpb.TargetStatus, e *scpb.TemporaryIndex) bool {
> 			return e.IndexID == indexID+1
> ```

Could we also lookup by checking `e.SourceIndexID`? That should be set for the index this temporary index was created for. 
___
*[`pkg/sql/regions/region_util.go` line 0 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141386#-OJdntGG29yinNZ0Oo7d:-OJdntGH7zh_K8kO6MFu:b-4dx8en) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d6d787c7f43c40fd0b1825a93cfd0141f5188c8/pkg/sql/regions/region_util.go#L0)):*
Is this just a straight copy and rename from the pkg/sql/region_util.go? Or were there any logic changes?


<!-- Sent from Reviewable.io -->

annrpom committed:
sql: move some MR zone config logic to new package

Mechanical changes (copy/paste) to avoid some cyclic
deps once these are needed in DSC.

Epic: none
Informs: #80545

Release note: None

annrpom committed:
schemachanger: port over MR zone config helpers

Some precursor work to remove RBR fallbacks for ADD COLUMN.

Epic: CRDB-31282
Informs: #80545

Release note: None

annrpom commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go:
```diff
@@ -1333,3 +1350,174 @@ func constructSideEffectPartitionElem(
+	if newZoneConfig.IsSubzonePlaceholder() && len(newZoneConfig.Subzones) == 0 {
+		return nil
+	}
+	tzc := &scpb.TableZoneConfig{TableID: tableID, ZoneConfig: &newZoneConfig}
```
done
<!-- Reviewable comment -OJe5yVxEl25mUW9rG00:bb7f2mf -->
<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141386)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)

___
*[`pkg/sql/regions/region_util.go` line  at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141386#-OJdntGG29yinNZ0Oo7d:-OJe5hii1rs3KmvM44Gg:bra8utr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d6d787c7f43c40fd0b1825a93cfd0141f5188c8/pkg/sql/regions/region_util.go#L)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Is this just a straight copy and rename from the pkg/sql/region_util.go? Or were there any logic changes?
</blockquote></details>

straight copy and rename! done to resolve some dependency cycles; i just moved these to a new commit in this PR
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 1260 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141386#-OJdp1Kd6r9r-CPyfF2X:-OJe5sSeAkyVJuBQ3ZJy:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d6d787c7f43c40fd0b1825a93cfd0141f5188c8/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L1260)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Could we also lookup by checking `e.SourceIndexID`? That should be set for the index this temporary index was created for.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> Reviewed 2 of 2 files at r2, 1 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141386)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 1394 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141386#-OJeFy5m9B2bEIV1Wrx3:-OJeFy5n4OzohHrtfikr:bpz7555) ([raw file](https://github.com/cockroachdb/cockroach/blob/db829f32f8630c1f653e970633dbe175beb55198/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L1394)):*
> ```Go
> // ApplyZoneConfigForMultiRegionTable applies zone config settings based
> // on the options provided and adds the scpb.TableZoneConfig to our builder.
> func ApplyZoneConfigForMultiRegionTable(
> ```

nit: Maybe this will make sense in a follow-on PR, but will we need to export this function out of this package?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 1482 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141386#-OJeJ2slAgabuC60Y13t:-OJeJ2slAgabuC60Y13u:bfxvyg0) ([raw file](https://github.com/cockroachdb/cockroach/blob/db829f32f8630c1f653e970633dbe175beb55198/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L1482)):*
> ```Go
> 		newZoneConfig.SubzoneSpans = nil
> 	}
> 	if newZoneConfig.IsSubzonePlaceholder() && len(newZoneConfig.Subzones) == 0 {
> ```

Is the placeholder subzone never materialized in the scpb elements?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 1486 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141386#-OJeJHwf3CieFpmMJ_Uh:-OJeJHwf3CieFpmMJ_Ui:b-shtnqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/db829f32f8630c1f653e970633dbe175beb55198/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L1486)):*
> ```Go
> 	}
> 	tzc := &scpb.TableZoneConfig{TableID: tableID, ZoneConfig: &newZoneConfig, SeqNum: 1}
> 	b.Add(tzc)
> ```

Do we need to delete the old table zone config? Or will that be handled by the caller in another PR.


<!-- Sent from Reviewable.io -->

annrpom reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141386)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 1394 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141386#-OJeFy5m9B2bEIV1Wrx3:-OJeOz8A7N0XTD6Yu9gs:b5l9k3f) ([raw file](https://github.com/cockroachdb/cockroach/blob/db829f32f8630c1f653e970633dbe175beb55198/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L1394)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

nit: Maybe this will make sense in a follow-on PR, but will we need to export this function out of this package?
</blockquote></details>

we will not! ty for catching -- my other PR actually just un-exports this back 
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 1482 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141386#-OJeJ2slAgabuC60Y13t:-OJeQmyu1A8Y86My1dn0:bgxb2dw) ([raw file](https://github.com/cockroachdb/cockroach/blob/db829f32f8630c1f653e970633dbe175beb55198/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L1482)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Is the placeholder subzone never materialized in the scpb elements?
</blockquote></details>

i believe this logic is specifying a no-op situation -- since a zone config that is a subzone placeholder means it's just there for subzones (indexes/partitions) to specify their configs and if there are no subzones, we don't actually need to write any zone config

lmk if i understand your question incorrectly
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 1486 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141386#-OJeJHwf3CieFpmMJ_Uh:-OK-H14F3Xy1X26SmaQN:bgnem6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/db829f32f8630c1f653e970633dbe175beb55198/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L1486)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Do we need to delete the old table zone config? Or will that be handled by the caller in another PR.
</blockquote></details>

we should be fine leaving the old tbl zc since we will always ref the tbl element with the greatest seqNum, but this also makes me realize this seqNum logic should be getting the latest seqNum+1 instead of just using 1... let me fix that (ty for the catch)


<!-- Sent from Reviewable.io -->

spilchen reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141386#-:-OK2wrIsEVRLClwU7OSd:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go` line 1482 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/141386#-OJeJ2slAgabuC60Y13t:-OK2vgnr68bCgVMTnZWx:bnljxxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/db829f32f8630c1f653e970633dbe175beb55198/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/zone_config_helpers.go#L1482)):*
<details><summary><i>Previously, annrpom (annie pompa) wrote…</i></summary><blockquote>

i believe this logic is specifying a no-op situation -- since a zone config that is a subzone placeholder means it's just there for subzones (indexes/partitions) to specify their configs and if there are no subzones, we don't actually need to write any zone config

lmk if i understand your question incorrectly
</blockquote></details>

Yep, makes sense. Thanks.


<!-- Sent from Reviewable.io -->

annrpom commented:
TFTR! ('-')7

bors r+
